src/reconstruction.jl

```diff
@@ -0,0 +1,214 @@
+
+#####################################################################################
+#                 Spatio Temporal Delay Embedding Reconstruction                    #
+#####################################################################################
+"""
+    AbstractSpatialEmbedding <: AbstractEmbedding
+Super-type of spatiotemporal embedding methods.
+Use `subtypes(AbstractSpatialEmbedding)` for available methods.
+"""
+abstract type AbstractSpatialEmbedding{T,Φ,BC,X} <: AbstractEmbedding end
+const ASE = AbstractSpatialEmbedding
+"""
+    AbstractBoundaryCondition
+Super-type of boundary conditions for [`SpatioTemporalEmbedding`](@ref).
+Use `subtypes(AbstractBoundaryCondition)` for available methods.
+"""
+abstract type AbstractBoundaryCondition end
+
+"""
+	ConstantBoundary(C)
+Constant boundary condition type. Enforces constant boundary conditions
+when passed to [`SpatioTemporalEmbedding`](@ref)
+by filling missing out-of-bounds values in the reconstruction with
+parameter `C`.
+"""
+struct ConstantBoundary{T} <: AbstractBoundaryCondition
+    C::T
+end
+
+"""
+	PeriodicBoundary
+Periodic boundary condition struct. Enforces periodic boundary conditions
+when passed to [`SpatioTemporalEmbedding`](@ref) in the reconstruction.
+"""
+struct PeriodicBoundary    <: AbstractBoundaryCondition end
+
+
+"""
+	Region{Φ}
+Internal struct for efficiently keeping track of region far from boundaries of field.
+Used to speed up reconstruction process.
+"""
+struct Region{Φ}
+	mini::NTuple{Φ,Int}
+	maxi::NTuple{Φ,Int}
+end
+
+Base.length(r::Region{Φ}) where Φ = prod(r.maxi .- r.mini .+1)
+function Base.in(idx, r::Region{Φ}) where Φ
+	for φ=1:Φ
+		r.mini[φ] <= idx[φ] <= r.maxi[φ] || return false
+ 	end
+ 	return true
+end
+Base.CartesianIndices(r::Region{Φ}) where Φ =
+	 CartesianIndices{Φ,NTuple{Φ,UnitRange{Int}}}(([r.mini[φ]:r.maxi[φ] for φ=1:Φ]...,))
+
+
+function inner_region(βs::Vector{CartesianIndex{Φ}}, fsize) where Φ
+	mini = Int[]
+	maxi = Int[]
+	for φ = 1:Φ
+		js = map(β -> β[φ], βs) # jth entries
+		mi,ma = extrema(js)
+		push!(mini, 1 - min(mi, 0))
+		push!(maxi,fsize[φ] - max(ma, 0))
+	end
+	return Region{Φ}((mini...,), (maxi...,))
+end
+
+function project_inside(α::CartesianIndex{Φ}, r::Region{Φ}) where Φ
+	CartesianIndex(mod.(α.I .-1, r.maxi).+1)
+end
+
+"""
+	 SpatioTemporalEmbedding{T,Φ,BC,X} <: AbstractSpatialEmbedding{T,Φ,BC,X} → `embedding`
+A spatio temporal delay coordinates structure to be used as a functor.
+
+	embedding(rvec, s, t, α)
+Operates inplace on `rvec` and reconstructs vector from spatial timeseries `s` at
+timestep `t` and cartesian index `α`.
+
+## Constructors
+The structure can be created directly by calling
+
+	SpatioTemporalEmbedding{T,Φ,BC,X}(τ,β,fsize)
+where `T` is the `eltype` of the timeseries, `Φ` the spatial dimension of the system,
+`BC` the boundary condition type and `X` the length of reconstructed vectors.
+Arguments `τ` and `β` are Vectors of `Int` and `CartesianIndex` that contain
```

Comment:
I don't understand what `β` is. Is it the relative indices of the "cell"? If yes, then this is nice and very general!

But maybe this construct is too advanced to be exposed here, but instead in some "advanced" section of the docs?